src/mono/mono/mini/interp/interp.c

```diff
@@ -3897,6 +3926,8 @@ interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClauseArgs
 				frame->state.ip = ip;
 				error_init_reuse (error);
 				do_jit_call (context, (stackval*)(locals + return_offset), (stackval*)(locals + call_args_offset), frame, cmethod, error);
+				if (context->pending_unwind)
+					return;
```

Comment:
We shouldn't simply return from `interp_exec_method`. Interpreter frames should be exited one by one. By simply returning we fail to restore the `context->data_stack`. `pending_unwind` should follow the same behavior as `has_resume_state` when it comes to exiting interpreter frames, they should probably be partly unified.